File: Java/src/main/java/Lists/SingleRowKeyboard/SingleRowKeyboard.java
package Lists.SingleRowKeyboard;

import java.util.LinkedHashSet;

public class SingleRowKeyboard {

    public static void main(String [] args) {
        String keyboard = "abcdefghijklmnopqrstuvwxyz";
        String word = "cba";

//        String keyboard = "pqrstuvwxyzabcdefghijklmno";
//        String word = "leetcode";

        System.out.println("Calculcated Time: " + calculateTime(keyboard, word));
    }

    public static int calculateTime(String keyboard, String word) {
        int total = 0;
        LinkedHashSet<Character> set = new LinkedHashSet<Character>();

        int index = 0;  // Need to track current Index

        // Traverse Loop
        for (int i=0; i < word.length(); i++) {
            char letter = word.charAt(i);

            // Track the targetLetter index
            int targetLetterIndex = keyboard.indexOf(letter);

            // Find the difference
            int diff = Math.abs(targetLetterIndex - index);

            // Add diff to total
            total += diff;

            // Set current index to targetedIndex
            index = targetLetterIndex;
        }

        return total;
    }



}

File: Python/project/src/Strings/SingleRowKeyboard/SingleRowKeyboard.py
def main():
    # keyboard = "abcdefghijklmnopqrstuvwxyz"
    # word = "cba"

    keyboard = "pqrstuvwxyzabcdefghijklmno"
    word = "leetcode"

    solution = Solution()
    output = solution.calculateTime(keyboard, word)
    print("Time: " + str(output))

class Solution(object):
    def calculateTime(self, keyboard, word):
        print("Keyboard: " + keyboard + "     word: " + word)

        # Step1 - Create output variable
        output = 0

        # Step2 - Track the prev variable
        prevIndex = 0

        # Step3 - Traverse each letter in word
        for letter in word:

            # Step4 - Find the index of the letter
            index = keyboard.find(letter)

            # Step5 - Calculate the difference and add to the output
            output += abs(prevIndex - index)

            # Step6 - Set the prev index
            prevIndex = index

        return output

if __name__ == '__main__':
    main()


'''
Notes
- Easy
- We use an absolute value to make sure diff is always positive


Solution
- Traverse each letter in word
- Find the index of the current word
- Calculate the difference between the prevWord and current Word  & Add difference to output variable
- Set the prevIndex to the current one
- Repeat until all words are covered

'''